scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/scala/uk/gov/nationalarchives/ingestvalidategenericingestinputs/Lambda.scala

```diff
@@ -0,0 +1,261 @@
+                      ("propertyWithError", Json.fromString(missingPropertyError.propertyWithError)),
+                      ("errorMessage", Json.fromString(missingPropertyError.errorMessage))
+                    )
+                  case validationError: ValidationError =>
+                    Json.obj(
+                      ("errorType", Json.fromString(validationError.getClass.getSimpleName)),
+                      ("errorMessage", Json.fromString(validationError.errorMessage))
+                    )
+                }.asJson
+            }
+        }
+      }.asJson
+
+  override def dependencies(config: Config): IO[Dependencies] =
+    IO(Dependencies(DAS3Client[IO]()))
+
+  override def handler: (
+      Input,
+      Config,
+      Dependencies
+  ) => IO[StateOutput] = (input, config, dependencies) =>
+    for {
+      log <- IO(log(Map("batchRef" -> input.batchId)))
+      _ <- log(s"Processing batchRef ${input.batchId}")
+      s3Client = dependencies.s3
+      bucket = input.metadataPackage.getHost
+      key = input.metadataPackage.getPath.drop(1)
+      metadataJson <- parseFileFromS3(s3Client, bucket, key)
+      _ <- log("Retrieving metadata.json from s3 bucket")
+
+      minimumAssetsAndFilesErrors <- MetadataJsonSchemaValidator.checkJsonForMinimumObjects(metadataJson)
+      _ <- log("Checking that JSON has at least one Asset and one File")
+
+      atLeastOneEntryWithSeriesAndNullParentErrors <- MetadataJsonSchemaValidator.checkJsonForAtLeastOneEntryWithSeriesAndNullParent(metadataJson)
+      _ <- log("Checking that JSON has at least one entry that has both a series and a null parent")
+
+      validatedEntries <- validateAgainstSchema(metadataJson) // all entries validated except for where the 'type' could not be found
+      _ <- log("Checking that each entry type in JSON matches a schema")
+
+      valueValidator = new MetadataJsonValueValidator
+      fileEntries = validatedEntries("File")
+      validatedLocations <- valueValidator.checkFileIsInCorrectS3Location(s3Client, fileEntries)
+
+      fileEntriesWithValidatedLocation = fileEntries.zip(validatedLocations).map { case (entries, locationNel) => entries + (location -> locationNel) }
+
+      (fileEntriesWithValidatedExtensions, metadataFileIds) = {
+        val idsOfMetadataFiles = validatedEntries("Asset").flatMap { entry =>
+          entry.collect {
+            case (fieldName, validatedValue) if fieldName == "originalMetadataFiles" =>
+              val listOfValues = validatedValue.getOrElse(Arr()).arr.toList
+              listOfValues.map(_.str)
+          }.flatten
+        }
+        val (metadataEntries, nonMetadataEntries) = fileEntriesWithValidatedLocation.partition { fileEntry =>
+          val fileEntryId = fileEntry(id).getOrElse(Str("idFieldHasErrorInIt")).str
+          idsOfMetadataFiles.contains(fileEntryId)
+        }
+        val metadataEntriesWithValidatedExtensions = valueValidator.checkFileNamesHaveExtensions(metadataEntries)
```

Comment:
This is only checking that metadata files have extensions; shouldn't we check this?

Reply:
I don't think it's necessary. So long as the right path is passed through, it doesn't matter what things are called.

Reply:
But Kurtis said that we need these to have extensions for Preservica/DROID reasons

Reply:
Maybe, but it won't break the ingest. If we get junk data through and it ends up a bit weird in Preservica then so be it. This validator is just to prevent the ingest breaking.

Reply:
Ah, I see. Ok, I'll remove it. Although, I just had a thought, instead of confirming that it has an extension, is it worth confirming that it has a `.json` extension because it not being a json would break the mapper Lambda? I know the file extension being `.json` doesn't necessarily prove that it is a JSON but it will at least capture a mistake.